pkg/restic/repository_manager.go

```diff
@@ -79,20 +82,19 @@ type RestorerFactory interface {
 }
 
 type repositoryManager struct {
-	namespace                    string
-	veleroClient                 clientset.Interface
-	secretsLister                corev1listers.SecretLister
-	repoLister                   velerov1listers.ResticRepositoryLister
-	repoInformerSynced           cache.InformerSynced
-	backupLocationLister         velerov1listers.BackupStorageLocationLister
-	backupLocationInformerSynced cache.InformerSynced
-	log                          logrus.FieldLogger
-	repoLocker                   *repoLocker
-	repoEnsurer                  *repositoryEnsurer
-	fileSystem                   filesystem.Interface
-	ctx                          context.Context
-	pvcClient                    corev1client.PersistentVolumeClaimsGetter
-	pvClient                     corev1client.PersistentVolumesGetter
+	namespace          string
+	veleroClient       clientset.Interface
+	secretsLister      corev1listers.SecretLister
+	repoLister         velerov1listers.ResticRepositoryLister
+	repoInformerSynced cache.InformerSynced
+	kbCache            kbcache.Cache
```

Comment:
That is correct. I think I had added this before I understood that get/list come from cache when using the client, then forgot to change it. Changing it now.

Reply:
Oh, actually, no. I used it so I would have access to `WaitForCacheSync`, which the client doesn't provide. I think we need this check? Please let me know.

https://github.com/vmware-tanzu/velero/blob/84698c88ff22229434fa3da6d8cbd99b9a4c7def/pkg/restic/repository_manager.go#L273

Reply:
so this is very related to https://github.com/vmware-tanzu/velero/pull/2561#discussion_r438242022, and I'm not 100% confident I understand the controller-runtime client behavior here. My sense was that it combines a ~lister/cache and an actual API client behind the scenes, so it'll return an object from the cache if it can, else it'll hit the API.

If that's the case, then we *may* not need to wait for the cache sync here, since we can use the client regardless. But again, i'm not totally sure I've got the mechanics of the new client down. Perhaps @ashish-amarnath can weigh in here.

Reply:
If you are using something that looks like this to create a client:

https://github.com/kubernetes-sigs/cluster-api/blob/714b0a97484465a05a67d8ccd1b64aa388b9fac2/util/util.go#L636-L646

Then any calls to `client.Get()` will not return any data until after the cache is synced.

Additionally, it will only return data from the cache. It does not fall back to doing a live read against the apiserver.

Does this help? Happy to hop on a call if you need it.

Reply:
I don't have all the context around this change, but I'll try to weigh in :) Internally `controller-runtime`'s cache will `WaitForCacheSync` when it tries to access any of the informers it is managing, so I'm not sure why you'd have to call it directly.

Reply:
I think that clarifies the question, @ncdc, thank you!

Reply:
> My sense was that it combines a ~lister/cache and an actual API client behind the scenes, so it'll return an object from the cache if it can, else it'll hit the API.

@skriss that's not quite my understanding.
If you use the default _Client_ setup by the controller-runtime Manger, it will delegate to the informer cache for typed resource reads and pass-through to the API Server for unstructured resource reads. For writes it will always pass-through to the API Server.

That is wired up [here](https://github.com/kubernetes-sigs/controller-runtime/blob/32b4434331e72fed344f71d278982827e08db2a8/pkg/manager/manager.go#L347).

If you use the _Cache_, you should be hitting all informers all the time.

Reply:
Note, in Cluster API, we explicitly made the choice to use a cache for all resources (typed and unstructured).

Reply:
thanks @shomron and @ncdc, that's helpful clarification

Reply:
Yes, what @shomron said is what I learned too. In other words, if the cache isn't started for example, the get/list will fail. Ways around that are, specifically pass an unstructured obj, or call the `GetAPIReader()`.

Reply:
If you use the controller-runtime `Manager`, it auto-starts all the caches for all the kinds you care about (have reconcilers for) automatically, so you generally don't need to worry about caches not being started in that setup.

Reply:
Hm. That's a good point. At this point there is no reconciler and no manager, we only have the resource.

Reply:
No actually, that's wrong, we do have a manager, just not any reconciler.